src/ddsrt/src/events/darwin/events_darwin.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright(c) 2006 to 2020 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/event.h>
+#include <sys/time.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "dds/ddsrt/events/darwin.h"
+#include "dds/ddsrt/events.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/sync.h"
+
+#define EVENTS_CONTAINER_DELTA 8
+
+ /**
+ * @brief Darwin (Apple) implementation of ddsrt_event_queue.
+ *
+ * This implementation uses a kqueue for monitoring a set of filedescriptors for events.
+ * Using the kevent call, the kernel can be told to add/modify fd's on its list for monitoring
+ * or to wait for events on the monitored fd's. Interrupts of waits are done through writes
+ * to an internal pipe.
+ */
+struct ddsrt_event_queue {
+  ddsrt_event_t**         events;   /**< container for triggered events*/
+  size_t                  nevents;  /**< number of triggered events stored*/
+  size_t                  cevents;  /**< capacity of triggered events stored*/
+  size_t                  ievents;  /**< current iterator for getting the next triggered event*/
+  int                     kq;       /**< kevent polling instance*/
+  struct kevent*          kevents;  /**< array which kevent uses to write back to, has identical size as this->events*/
+  ddsrt_mutex_t           lock;     /**< for keeping adds/deletes from occurring simultaneously */
+#if !defined(LWIP_SOCKET)
+  ddsrt_socket_t          interrupt[2]; /**< pipe for interrupting waits*/
+#endif
+};
+
+/**
+* @brief Initializes an event queue.
+*
+* Will set the counters to 0 and create the containers for triggers and additional necessary ones.
+*
+* @param[out] queue The queue to initialize.
+*
+* @returns DDS_RETCODE_OK if everything went OK.
+*/
+static dds_return_t ddsrt_event_queue_init(ddsrt_event_queue_t* queue) {
+  queue->nevents = 0;
+  queue->cevents = 8;
+  queue->ievents = 0;
+  ddsrt_mutex_init(&queue->lock);
+
+  /*create kevent polling instance */
+  queue->kq = kqueue();
+  assert(queue->kq != -1);
+  assert(fcntl(queue->kq, F_SETFD, fcntl(queue->kq, F_GETFD) | FD_CLOEXEC) != -1);
+
+  queue->events = ddsrt_malloc(sizeof(ddsrt_event_t*) * queue->cevents);
+  assert(queue->events);
+
+#if !defined(LWIP_SOCKET)
+  if (-1 == pipe(queue->interrupt))
+    return DDS_RETCODE_ERROR;
+  /*register interrupt event*/
```

Comment:
It would probably be a good idea to make the pipe non-blocking and ignore `EAGAIN`: the point is to trigger it, a full buffer should not block the the trigger operation and spurious triggers are ok.

Reply:
This has now been implemented.